### PR TITLE
kafka-3739:Add no-arg constructor for library provided serdes

### PR DESCRIPTION
**Pull request by huxihx**
Add default constructor for library provided serdes

### CODE REVIEW

**Comment by huxihx** [comment]
@guozhangwang Added default constructors for WindowedSerde/ChangedSerde, but not for JsonPOJOSerde since it is in the example package. Please review the PR. Thanks.

**Comment by guozhangwang** [comment]
Thanks for the patch, a couple of meta comments:

1. `ChangeSerde` is used internally only and not exposed to users, so we do not need to add no-arg constructor to it.

2. For `WindowedSerde`, inner serde is required, so if it is not passed in constructor, which is possible, we need to construct it in `configure` function, which will be auto-triggered when the instance is created by reflection. You can take a look at `StringSerde` class as a semi-related reference. Personally I'm thinking the inner serde could be again constructed via reflection from a passed class type string, and then assigned to `inner`.

**Comment by huxihx** [comment]
@guozhangwang I am not sure I fully understand what you mean, but what do you think if I add the code like this:
In WindowedSerializer.configure method, 
`if (inner == null) {`
`    inner = new ProducerConfig(configs).getConfiguredInstance(isKey ? "key.serializer" : "value.serializer", Serializer.class);`
`    inner.configure(configs, isKey);`
`}`
However, seems that we have to modify the access level for ProducerConfig init to 'public', do you live with that?

Besides, that's the unit test code:
`@Test
    public void testWindowedSerdeNoArgConstructors() {

        WindowedSerializer<StringSerializer> windowedSerializer = new WindowedSerializer<>();

        Map<String, String> props = new HashMap<>();

        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "host:1");

        props.put(StreamsConfig.APPLICATION_ID_CONFIG, "appId");

        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        windowedSerializer.configure(props, true);

        Serializer<?> inner = windowedSerializer.getInnerSer();

        assertNotNull("Inner serializer should be not null", inner);

        assertTrue("inner serializer should be with StringSer", inner instanceof StringSerializer);

    }`

Forgot to say, I created a method named 'getInnerSer' in WindowedSerializer to get `inner` only for testing.

**Comment by guozhangwang** [comment]
The point is, in either way how the `WindowedSer/Deser` was created (either explicitly internal of Streams, or via reflection from the config class name), the inner serde should always be set after it is constructed since otherwise it cannot be used.

I think it is better to have the config name sth. like `key[value].serializer[deserializer].inner.class`? 

And instead of creating a `ProducerConfig` object just to call its `getConfiguredInstance` I think we can just inline this function inside `configure`.

**Comment by huxihx** [comment]
@guozhangwang  Thanks for the comments. Added two configs in StreamsConfig: "key.serializer.inner.class" and "value.serializer.inner.class"
And it's not easy to inline the function 'AbstractConfig.getConfiguredInstance' since it internally invokes other methods and references some internal states

Please have a review. If you think is okay, I will work on the deserializer counterparts. Thanks.

**Comment by guozhangwang** [comment]
Code checkstyle failed:

```
scala2.12/streams/src/test/java/org/apache/kafka/streams/kstream/internals/WindowedStreamPartitionerTest.java:31: Using the '.*' form of import should be avoided - java.util.*.
```

**Comment by guozhangwang** [comment]
Also we do not need to add these two configs into StreamsConfig, just as we did not add `key.serializer.encoding` used in StringSerializer in ProducerConfig. Instead, we should just mention them in the javadoc above as StringSerializer.

**Comment by huxihx** [comment]
@guozhangwang addressed your comments including inlining AbstractConfig.getConfiguredInstance and refining some Javadoc. What about this time? If okay, will work on the deserializer parts. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java [hunk 1]
```diff
@@ -17,13 +17,19 @@
...
 import org.apache.kafka.common.serialization.Serializer;
-import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.streams.kstream.Windowed;
 
 import java.nio.ByteBuffer;
 import java.util.Map;
 
+/**
+ *  The inner serializer class must be specified by setting the property key.serializer.inner.class,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java]: `must` -> `can be specified ... if the no-arg constructor is called and hence it is not passed during initialization`.

**Review by guozhangwang — COMMENTED** [review]
LGTM overall. Please note there are still some checkstyle failures. Pleas use `./gradlew jar` and `./gradlew test` to make sure that checkstyle and unit tests pass.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java (hunk 1 above)
Comment (amethystic) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java]: I did not encounter any checksytle failures. Could you kindly point out which files failed to pass this check?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java]: You can click on the below links named "Details" to see the failure message. Also try using `./gradlew checkstyleMain checkstyleTest` to check locally.

**Comment by huxihx** [comment]
@guozhangwang already added the code for WindowedDeserializer and passed the style check. Please review again. Thanks.

**Comment by huxihx** [comment]
@guozhangwang Please review the PR again. Thanks.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java [hunk 2]
```diff
@@ -17,25 +17,53 @@
...
-        // do nothing
+        if (inner == null) {
+            String propertyName = isKey ? "key.deserializer.inner.class" : "value.deserializer.inner.class";
+            Object innerDeserializerClass = configs.get(propertyName);
+            propertyName = (innerDeserializerClass == null) ? "deserializer.inner.class" : propertyName;
+            String value = null;
+            try {
+                value = (String) configs.get(propertyName);
+                Object o = Utils.newInstance(value, Deserializer.class);
+                if (!Deserializer.class.isInstance(o)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: Will this ever happen? I think `Utils.newInstance` is guaranteed to give you an instance of the passed in `class` type. Ditto below.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java [hunk 3]
```diff
@@ -33,9 +42,28 @@ public WindowedSerializer(Serializer<T> inner) {
...
+        if (inner == null) {
+            String propertyName = isKey ? "key.serializer.inner.class" : "value.serializer.inner.class";
+            Object innerSerializerClass = configs.get(propertyName);
+            propertyName = (innerSerializerClass == null) ? "serializer.inner.class" : propertyName;
+            String value = null;
+            try {
+                value = (String) configs.get(propertyName);
+                Object o = Utils.newInstance(value, Serializer.class);
+                if (!Serializer.class.isInstance(o)) {
+                    throw new KafkaException(value + " is not an instance of Serializer");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedSerializer.java]: Ditto above.

**Review by guozhangwang — COMMENTED** [review]
One minor comment, otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java (hunk 2 above)
Comment (huxihx) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: Agreed. However, the check is introduced by AbstractConfig.getConfiguredInstance. Do you think we should just keep the same as it or remove this unnecessary check only for two classes?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedDeserializer.java]: I looked at it, and I think the scenario is a bit different in  `public <T> T getConfiguredInstance(String key, Class<T> t)` as in here, where `Utils.newInstance(Class<T> c)` is called and the passed `Class<?>` is directly from the config, where in this case the config value is a `String` and we directly use that in `Utils.newInstance`.

**Comment by huxihx** [comment]
@guozhangwang Already removed the unnecessary class type check in WindowedSerde configure methods. Please kindly review again. Thanks.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
Merged to trunk. Thanks @amethystic and congrats for your first commit!

**Comment by huxihx** [comment]
Thanks for your patience and your time:)

**Pull request closed without merging**